### ISSUE TITLE
Sorting before Windows functions - example from Polars book

### ISSUE THREAD

**Issue by lucazanna** [pola-rs/polars#8958]
### Research

- [X] I have searched the above polars tags on Stack Overflow for similar questions.

- [ ] I have asked my usage related question on [Stack Overflow](https://stackoverflow.com).

### Question about Polars

I am reading the example in the Polars guide about window functions:
https://pola-rs.github.io/polars-book/user-guide/expressions/window/#more-examples

The example mentions at the end 'for this operation to make sense, it is important that the columns used in over([..]) are sorted!'

Can anyone explain the reason for sorting before calling the select context?
To my understanding, window functions run calculations within the group indicated by the `over` statement. So it should not matter if the rows belonging to a certain group are not contiguous.

I just ran a check for the example for both:
`sorted_ = df.sort("Type 1").select( ...`
or 
`non_sorted_ = df.select( ...`

and the result is the same (except for the order of the rows, which might not matter depending on the use case)

**Comment by ritchie46** [comment]
This whole section with list and flatten doesn't make sense anymore. You are right, the window function deal with that itself.

**Comment by lucazanna** [comment]
Hi @ritchie46 

can you share why it does not make sense anymore ?
Except from the sorting, the rest seems to make sense

i can submit a PR to change it

**Comment by ritchie46** [comment]
First calling `implode` and then calling `flatten` is convoluted. Is similar to doing a shuffle followed by a sort.

**Comment by lucazanna** [comment]
Hi @ritchie46 ,

I understand and yes it took me some time to wrap my head around this. Are you planning to change how `flatten` or `implode` work?

With the current behaviour, I did not find a way to obtain the same result without using `implode` + `flatten`.

I also found something strange with `flatten`, I will open a separate issue for it

**Comment by cmdlineluser** [comment]
`.arr.head` can be used:
```python
(df.sort("Type 1")
   .select(
      pl.implode("Type 1").arr.head(3).over("Type 1"),
      pl.col("Name").sort_by("Speed") .implode().arr.head(3).over("Type 1").alias("fastest/group"),
      pl.col("Name").sort_by("Attack").implode().arr.head(3).over("Type 1").alias("strongest/group"),
      pl.col("Name").sort_by("Name")  .implode().arr.head(3).over("Type 1").alias("sorted_by_alphabet")
   )
)
```
However, I get an exception if I then remove the `.sort`:
```
ComputeError: the length of the window expression did not match that of the group
> group: ("Grass")
> group length: 1
> output: 'shape: (3,)
Series: '' [str]
[
	"Grass"
	"Grass"
	"Grass"
]'

Error originated in expression: 'col("Type 1").list().slice([0, 3]).over([col("Type 1")])'
```
If I use `.sort(True)` - I get a single row result:
```
shape: (1, 4)
┌────────┬───────────────┬─────────────────┬────────────────────┐
│ Type 1 ┆ fastest/group ┆ strongest/group ┆ sorted_by_alphabet │
│ ---    ┆ ---           ┆ ---             ┆ ---                │
│ str    ┆ str           ┆ str             ┆ str                │
╞════════╪═══════════════╪═════════════════╪════════════════════╡
│ Grass  ┆ Bulbasaur     ┆ Bulbasaur       ┆ Bulbasaur          │
└────────┴───────────────┴─────────────────┴────────────────────┘
```

I know it's supposed to be an example for window expressions, but it seems more suited to `.groupby` instead of repeating the same `.implode().over()` for so many expressions:
```python
(df.groupby("Type 1")
   .agg(
      pl.col("Name").sort_by("Speed") .head(3).alias("fastest/group"),
      pl.col("Name").sort_by("Attack").head(3).alias("strongest/group"),
      pl.col("Name").sort_by("Name")  .head(3).alias("sorted_by_alphabet"),
   )
   .sort("Type 1")
   .explode(pl.exclude("Type 1"))
)
```

**Comment by lucazanna** [comment]
@cmdlineluser : you are right, it works with the `.arr`

One thing that is unclear to me :

Without the arr.head we get a column of lists:
```python
df.sort("Type 1")
   .select(
      pl.implode("Type 1").over("Type 1"),
   )
)

┌───────────────────────────────┐
│ Type 1                        │
│ ---                           │
│ list[str]                     │
╞═══════════════════════════════╡
│ ["Bug", "Bug", … "Bug"]       │
│ ["Bug", "Bug", … "Bug"]       │
│ ["Bug", "Bug", … "Bug"]       │
│ ["Bug", "Bug", … "Bug"]       │
│ …                             │
│ ["Water", "Water", … "Water"] │
│ ["Water", "Water", … "Water"] │
│ ["Water", "Water", … "Water"] │
│ ["Water", "Water", … "Water"] │
└───────────────────────────────┘
```
with arr.head(3) we get a reordered column of type Utf8, with each group only once

```python
print(df.sort("Type 1")
   .select(
      pl.implode("Type 1").arr.head(3).over("Type 1"),
   )
)

┌────────┐
│ Type 1 │
│ ---    │
│ str    │
╞════════╡
│ Bug    │
│ Bug    │
│ Bug    │
│ Dragon │
│ …      │
│ Rock   │
│ Water  │
│ Water  │
│ Water  │
└────────┘
```
I would expect arr.head() to simply reduce the elements to 3 and not flatten + keep each group once.

I was thinking of keeping this issue only for the guide, so I opened a separate issue here: https://github.com/pola-rs/polars/issues/8965

**Comment by cmdlineluser** [comment]
> I would expect arr.head() to simply reduce the elements to 3 and not flatten + keep each group once.

Huh, yes - I never actually thought about the implicit flatten going on there - I'm now also a little confused as to why this happens.

**Comment by stinodego** [comment]
I believe the section in the book has been updated. Please open a new issue if there are still something wrong.
